csrc/scheduler/multi_matmul.cpp

```diff
@@ -0,0 +1,421 @@
+  //  shared mem cache of operands : acw_smem, bcw_smem (short for a/b
+  //  cache_write smem)
+  //
+  //  registers at shared memory load output : acr, bcr (short for a/b cache
+  //  read)
+  //
+  //  register tensor input to the actual mma op: ab, bb (short for a/b
+  //  broadcasted)
+  //
+  //  accumulator register: mma_result
+  //   - mma_result is MmaOp output if there is epilogue
+  //   - mma_result is dc (short for d cache) if there is no epilogue
+  //
+  //  result in global memory: d
+
+  // Currently the support is for a, b, c and d as fusion inputs/outputs
+  //  aka. no prolog fusion yet.
+  void defineOperandCaches() {
+    mma_results_.reserve(mma_ops_.size());
+    for (MmaOp* mma : mma_ops_) {
+      mma->setMacro(params_.mma_macro);
+
+      // Setup accumulator register.
+      mma_results_.push_back(mma->out()->as<TensorView>());
+    }
+
+    LoadStoreOpType load_op = params_.async_gmem_load_operands
+        ? LoadStoreOpType::CpAsync
+        : LoadStoreOpType::Set;
+
+    auto cacheOperandsToSmem = [&](const std::vector<TensorView*>& operands,
+                                   std::vector<TensorView*>& smem_operands,
+                                   int64_t vec_size) {
+      // Use cp.async as requested in scheduler params.
+      smem_operands.resize(operands.size(), nullptr);
+      for (size_t i : c10::irange(operands.size())) {
+        TensorView* operand = operands[i];
+        CacheOp cache_op = CacheOp::Unspecified;
+        if (params_.async_gmem_load_operands) {
+          int64_t vec_bytes = vec_size * dataTypeSize(operand->dtype());
+          NVF_CHECK(
+              vec_bytes == 4LL || vec_bytes == 8LL || vec_bytes == 16LL,
+              "Unsupported async vectorization size ",
+              vec_size,
+              " = ",
+              vec_bytes,
+              " bytes for operand ",
+              operand->toString(),
+              " which has data type ",
+              operand->dtype(),
+              ". Size must be 4, 8, or 16 bytes. ",
+              "MatmulParams::async_gmem_load_operands should be set to false in this case.");
+          cache_op = vec_bytes == 16LL ? CacheOp::Global : CacheOp::AllLevels;
+        };
+
+        NVF_ERROR(operand->uses().size() == 1);
+        smem_operands[i] = ir_utils::consumerTvsOf(operand).at(0);
+        smem_operands[i]->definition()->as<LoadStoreOp>()->setOpType(load_op);
+        smem_operands[i]->definition()->as<LoadStoreOp>()->setCacheOp(cache_op);
+        if (smem_operands[i]->uses().size() > 1) {
+          // There can be multiple uses for example if we have A @ B1 + A @ B2
+          // then A will be cached to smem then it might be loaded into two
+          // separate register buffers, one for each mma. Instead, we will load
+          // it once into registers then re-use the register buffer for both
+          // mmas.
+          cacheAfter(smem_operands[i]);
+        }
+        NVF_ERROR(smem_operands[i]->uses().size() == 1);
```

Comment:
Should we move this to `addSetsForCacheReads`? I believe the
```C++
if (auto ldst = dynamic_cast<LoadStoreOp*>(tv_smem->uses().at(0))) {
  ...
} else {
  ...
}
```
in `addSetsForCacheReads` will automatically handle the case where there are multiple uses, and only use a single ldmatrix?

Reply:
Removed here and added a uses().size() check in `addSetsForCacheReads`.